0001-kaa-protocol/README.md

```diff
@@ -93,40 +109,93 @@ Examples of extension specific paths are:
 
-### Request/response pattern
+After client authentication is done, the client may issue commands on behalf of endpoints.
+To do so, it needs an endpoint token.
+A token is sent along with each request as part of the [resource path](#Glossary).
+
+_Endpoint token_ is an opaque data blob that uniquely identifies an endpoint.
+Each endpoint has exactly one token assigned by the server.
+In the simplest form, a client always uses the same token, thus representing a single endpoint.
 
-Many extensions require request/response style communication, which is not supported by MQTT natively. That is usually overcome by introducing a separate topic name for responses.
+There are two possibilities for a client to get an endpoint token: it is either pre-provisioned into the application, or requested at run-time via an extension _(endpoint registration)_.
+Using an extension allows implementing different registration schemes as well as changing and allocating endpoint tokens dynamically.
 
-We introduce `/status` topic for response messages. In case original payload contains `packet_id`, response should be published in that topic.
+_Note: the server may require endpoint registration before using by specific client._
+_That way we can prevent endpoint stealing to some degree._
 
-_Note: that only applies for client-originating requests._
+_Note: we can assign endpoint tokens per client session._
+_That’s more secure and allows using short non-secure tokens._
+
+On the other hand, using a predefined token is a simpler but less secure alternative.
+In that case, the endpoint is manually added to the server, the server allocates a unique token and provides it to the user, the user embeds the token in the application.
 
 ## Open questions
 
 ### Topic name aliases
-
-It's good to keep topic names short. Maybe, we can introduce topic aliases, so user won't need to use full-length topic name.
+It's good to keep topic names short.
+Maybe we can introduce topic aliases, so user won't need to use full-length topic name.
 
 ### Endpoint migration
-An endpoint might migrate between different clients. How possible this use case is?
+An endpoint might migrate between different clients.
+
+#### UC1
+Endpoint is a wearable that communicates through stationary gateways.
+The endpoint thus can be seen through any of the gateways.
+Also, it can be seen through two or more gateways at the same time.
+
+#### UC2
+Endpoint has channel redundancy through two (or more) different gateways (or forward proxies).
+
+### Reporting errors
+In the current KP design it is not possible to respond with an error to a specific message.
+Errors may be:
+- appversion does not exit;
+- wrong extension instance name;
+- unauthorized (EP token not found);
+- etc.
+
+### QoS
+Which MQTT QoS levels compliant implementations must support?
+
+### Security
+- Which authentication combinations should the KP implementations support?
+- Is it possible to send unencrypted, yet signed data?
```

Comment:
Yeah - I was thinking about a different level that would also persist signatures in the DB to avoid further tampering. That's DCX-related, so never mind.